wren-ui/src/apollo/server/managers/dataSourceSchemaDetector.ts

```diff
@@ -0,0 +1,288 @@
+        }
+
+        // If there are any deleted or modified columns, we need to add them to the result.
+        if (deletedColumnChange.columns.length > 0) {
+          result[SchemaChangeType.DELETED_COLUMNS] = [
+            ...(result[SchemaChangeType.DELETED_COLUMNS] || []),
+            deletedColumnChange,
+          ];
+        }
+        if (modifiedColumnChange.columns.length > 0) {
+          result[SchemaChangeType.MODIFIED_COLUMNS] = [
+            ...(result[SchemaChangeType.MODIFIED_COLUMNS] || []),
+            modifiedColumnChange,
+          ];
+        }
+      }
+
+      return result;
+    }, {});
+
+    if (!isEmpty(diffSchema)) {
+      logger.debug('Diff Schema:', JSON.stringify(diffSchema));
+      logger.info('Data Source Schema has changed.');
+      return diffSchema as DataSourceSchemaChange;
+    }
+
+    logger.info('No changes in Data Source Schema.');
+    return null;
+  }
+
+  private async addSchemaChange(diffSchema: DataSourceSchemaChange) {
+    const getResolveState = (change) => (!!change ? false : null);
+
+    const lastSchemaChange =
+      await this.ctx.schemaChangeRepository.findLastSchemaChange(
+        this.projectId,
+      );
+    // If the schema change is the same as the last one, we don't need to create a new one.
+    const isNewSchemaChange =
+      JSON.stringify(lastSchemaChange?.change) !== JSON.stringify(diffSchema);
+
+    if (isNewSchemaChange) {
+      await this.ctx.schemaChangeRepository.createOne({
+        projectId: this.projectId,
+        change: diffSchema,
+        // Set the resolve to false if there are any changes. It will set resolve to true once the schema has been synced.
+        resolve: {
+          [SchemaChangeType.DELETED_TABLES]: getResolveState(
+            diffSchema[SchemaChangeType.DELETED_TABLES],
+          ),
+          [SchemaChangeType.DELETED_COLUMNS]: getResolveState(
+            diffSchema[SchemaChangeType.DELETED_COLUMNS],
+          ),
+          [SchemaChangeType.MODIFIED_COLUMNS]: getResolveState(
+            diffSchema[SchemaChangeType.MODIFIED_COLUMNS],
+          ),
+        },
+      });
+    }
+  }
+
+  private async getCurrentSchema(): Promise<DataSourceSchema[]> {
+    const models = await this.ctx.modelRepository.findAllBy({
+      projectId: this.projectId,
+    });
+    const modelIds = models.map((model) => model.id);
+    const modelColumns =
+      await this.ctx.modelColumnRepository.findColumnsByModelIds(modelIds);
+    const result = models.map((model) => {
+      return {
+        name: model.sourceTableName,
+        columns: modelColumns
+          .filter(
+            (column) => column.modelId === model.id && !column.isCalculated,
+          )
+          .map((column) => ({
+            name: column.sourceColumnName,
+            type: column.type,
+          })),
+      };
+    });
+    return result;
+  }
+
+  private async getLatestSchema(): Promise<DataSourceSchema[]> {
```

Comment:
I'll modify this method tmr since the Ibis integration PR has been merged.

Reply:
addressed in [db3e5b9](https://github.com/Canner/WrenAI/pull/404/commits/db3e5b9a866bf2bf564afaa329b31022478b0c96)